### PR TITLE
RPCA Seasonal Anomaly Detection

### PR DESCRIPTION
**Pull request by shouvikmani**
Anomaly detection based on the Robust Principal Component Analysis (RPCA) technique developed by Netflix. Particularly well-suited for detecting anomalies in seasonal data.

Article: http://techblog.netflix.com/2015/02/rad-outlier-detection-on-big-data.html

### CODE REVIEW

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    public String getResultMetricName() {
+        return RESULT_METRIC_NAME;
+    }
+
+    @Override
+    public double calculateAnomalyScore(double value) {
+        return Math.abs(value);
+    }
+
+    /**
```
Comment (saaja-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: Javadoc comment is not required here.  Simple multi-line comment should be used.

**Comment by saaja-sfdc** [comment]
@shouvikmani did the unit tests get dropped from the commit?

File: ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                    printCnt++;
+                    testOutput = false;
+                }
+            }
+        }
+        return testOutput;
+    }
+
+    @Test
+    public void testRSVD() {
```
Comment (afalko) [inline ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java]: Would be better to refactor this to a parameterized test (in junit parlance): https://github.com/junit-team/junit4/wiki/parameterized-tests

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics;
+
+/**
+ * RPCA code taken from Netflix Surus Project.
+ * Source: https://github.com/Netflix/Surus/blob/master/src/main/java/org/surus/math/RPCA.java
+ *
+ */
+public class RPCA {
+
+	private RealMatrix X;
```
Comment (afalko) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java]: You should make all of these class global vars `final`

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+ * Sources:
+ * (i) Article: http://techblog.netflix.com/2015/02/rad-outlier-detection-on-big-data.html
+ * (ii) Source code: https://github.com/Netflix/Surus
+ *
+ * Ex: ANOMALY_RPCA(-100d:-0d:foo:bar:sum, $7d), where $7d indicates that the length of
+ * a season of the metric is 7 days.
+ *
+ * @author  Shouvik Mani & Vignesh Muruganantham
+ */
+public class AnomalyDetectionRPCATransform extends AnomalyDetectionTransform {
```
Comment (afalko) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: Generally, you want to stay away from inheritance (extends and abstract classes) in favor of composition (implements).

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+    private int frequency;
```
Comment (afalko) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: All of these should be `final`

Comment (afalko) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: Because you're heavily relying on inheritance, it is difficult to make these immutable; which in turn increases the potential for bugs and makes it near impossible to do anything with concurrency later.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        }
+        double variance = squaredSum/values.length;
+        return Math.sqrt(variance);
+    }
+
+    private double[][] vectorToMatrix(double[] x, int rows, int cols) {
+        double[][] input2DArray = new double[rows][cols];
+        for (int n = 0; n < rows * cols; n++) {
+            int i = n % rows;
+            int j = (int) Math.floor(n / rows);
```
Comment (afalko) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: will rows ever come in `0`?

File: ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java (hunk 2 above)
Comment (afalko) [inline ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java]: Also if you name it parameterized test, it'll be much easier to see what you are inputting into all of the functions below....i.e. instead of seeing something like `VectorToMatrix(ts, 7, 9);` you'd see `VectorToMatrix(ts, rows, cols);` which is much more readable than seeing raw random numbers.

File: ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        // X:
+        double[] ts = new double[] {2.05407309078346,2.85886923211884,2.89728554463089,0.790480493540229,0.548595335194215,1.31367506547418,1.74407133897301,4.06071962679526,2.75651081738515,0.604658754735038,0.182607837501951,-1.262201503678,0.996560864201235,2.74637817075616,0.775004762296101,0.906823901472144,2.6839457174704,-0.0625841462071901,-1.09641353766956,0.00479165991036998,0.449351175604642,3.53152043857777,1.05206417605014,2.7864942275709,-0.691007430091048,-1.02038488026721,-1.35124486835257,0.0621976297222073,2.82421545538541,2.41312411015615,1.27711183784622,0.0988204592711682,1.50691474460298,0.272037685359444,1.9889742629239,3.33907184622517,3.68134545243902,0.751559686193563,0.679120355399832,0.428056866405207,0.351341204822829,1.33498418531095,3.04169869243666,1.22542459625713,1.35457091793328,0.567124649501233,-1.95560538335988,-1.09014280752067,1.80062291606412,0.588637569785287,1.89212604693897,1.38386740607786,0.356716316822486,-2.07161693692556,4,1.44451323393473,3.52551739267569,3.16481926426412,1.83839333727511,0.827646664705546,0.654351159135431,-0.00892931340717523,0.678082675364184};
+        double[][] X = VectorToMatrix(ts, 7, 9);
```
Comment (afalko) [inline ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java]: test cases for rows=0, cols=0

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+    @Override
+    public List<Metric> transform(List<Metric> metrics, List<String> constants) {
+        SystemAssert.requireArgument(metrics != null, "Cannot transform null metric/metrics");
+        SystemAssert.requireState(metrics.size() == 1, "Anomaly Detection Transform can only be used on one metric.");
+        SystemAssert.requireState(metrics.get(0) != null, "Anomaly Detection Transform cannot be used with a null " +
+                "metric.");
+        SystemAssert.requireState(constants.size() == 1, "Anomaly Detection RPCA Transform can only be used with " +
+                "one constant for the length of a season");
+
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: It will be better to combine the 3 asserts above? I still don't see a compelling reason of using asserts at runtime? Do these asserts have the same semantics as the asserts in test frameworks? In that case you wouldn't want to halt execution and instead do exception handling.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        //Create a sorted array of the metric's timestamps
+        Map<Long, String> completeDatapoints = metrics.get(0).getDatapoints();
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: Map is an interface, you might wan't to strongly type this to make it type safe. This is better for interfaces too as if most of your runtime is expecting to operate on a HashMap and instead you get a TreeMap then your code might break

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+ *
+ * Ex: ANOMALY_RPCA(-100d:-0d:foo:bar:sum, $7d), where $7d indicates that the length of
+ * a season of the metric is 7 days.
+ *
+ * @author  Shouvik Mani & Vignesh Muruganantham
+ */
+public class AnomalyDetectionRPCATransform extends AnomalyDetectionTransform {
+
+    private int frequency;
+    private Long[] timestamps;
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: Could use a better name as its a list of timestamps as it is a list of timestamps?

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+        SystemAssert.requireArgument(metrics != null, "Cannot transform null metric/metrics");
+        SystemAssert.requireState(metrics.size() == 1, "Anomaly Detection Transform can only be used on one metric.");
+        SystemAssert.requireState(metrics.get(0) != null, "Anomaly Detection Transform cannot be used with a null " +
+                "metric.");
+        SystemAssert.requireState(constants.size() == 1, "Anomaly Detection RPCA Transform can only be used with " +
+                "one constant for the length of a season");
+
+        //Create a sorted array of the metric's timestamps
+        Map<Long, String> completeDatapoints = metrics.get(0).getDatapoints();
+        timestamps = completeDatapoints.keySet().toArray(new Long[completeDatapoints.size()]);
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: You are using an uninitialized array here and using it in a global context. This has a private accessor in a public method. You should use getters and setters instead. There are several things that may be done in this context

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+        Arrays.sort(timestamps);
+
+        long seasonLengthInMilliseconds = getTimePeriodInSeconds(constants.get(0)) * 1000;
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: If this is a super method, its more readable to use the super identifier like super.getTimePeriodInSeconds. and constants.get(0) is also not readable. Usually constants are defined and accessed in a certain way. For e.g:
Constants.MAX_FREQUENCY.

File: ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        double[][] E = rsvd.getE().getData();
+        double[][] S = rsvd.getS().getData();
+        double[][] L = rsvd.getL().getData();
+
+        assertTrue(MatrixApproximatelyEquals(E_matrix_r, E, 0.0001));
+        assertTrue(MatrixApproximatelyEquals(S_matrix_r, S, 0.0001));
+        assertTrue(MatrixApproximatelyEquals(L_matrix_r, L, 0.0001));
+    }
+
+}
```
Comment (tuhinkanti) [inline ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java]: Test coverage looks very low. Also a unit test should do one thing and one thing well, what that means is if a test passes and fails it reliably communicates what failed. A unit test shouldn't be verifying multiple things since then a failure or success doesn't communicate a lot

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+	private RealMatrix L;
+	private RealMatrix S;
+	private RealMatrix E;
+
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java]: Better names than X,L, S and E. Even if they have mathematical significance you should use clean code principles

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/TransformFactory.java [hunk 15]
```diff
@@ -249,7 +251,8 @@ public Transform getTransform(String functionName) {
         GROUPBY("GROUPBY", "Creates groups of metrics based on some matching criteria and then performs the given aggregation."),
         ANOMALY_DENSITY("ANOMALY_DENSITY", "Calculates an anomaly score (0-100) for each value of the metric based on the probability density of each value with a Gaussian distribution."),
         ANOMALY_ZSCORE("ANOMALY_ZSCORE", "Calculates an anomaly score (0-100) for each value of the metric based on the z-score of each value with a Gaussian distribution."),
-        ANOMALY_KMEANS("ANOMALY_KMEANS", "Calculates an anomaly score (0-100) for each value of the metric based on a K-means clustering of the metric data.");
+        ANOMALY_KMEANS("ANOMALY_KMEANS", "Calculates an anomaly score (0-100) for each value of the metric based on a K-means clustering of the metric data."),
+        ANOMALY_RPCA("ANOMALY_RPCA", "Calculates an anomaly score (0-100) for each value of the metric based on the RPCA matrix decomposition algorithm.");
 
         private final String _name;
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/TransformFactory.java]: variable names with underscore isn't common in java

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+	private double spenalty;
+	
+	private static final int MAX_ITERS = 228;
+	
+	public RPCA(double[][] data, double lpenalty, double spenalty) {
+		this.X = MatrixUtils.createRealMatrix(data);
+		this.lpenalty = lpenalty;
+		this.spenalty = spenalty;
+		initMatrices();
+		computeRSVD();
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java]: Calling a method inside the constructor which isn't an initialization does not make sense. Why would you do that? Methods are invoked using an object. Again a class is a specification while an object is an instance of that specification.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java (hunk 4 above)
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: 👍

File: ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+
+        RPCA rsvd = new RPCA(X, 1, 1.4/3);
+
+        double[][] E = rsvd.getE().getData();
+        double[][] S = rsvd.getS().getData();
+        double[][] L = rsvd.getL().getData();
+
+        assertTrue(MatrixApproximatelyEquals(E_matrix_r, E, 0.0001));
+        assertTrue(MatrixApproximatelyEquals(S_matrix_r, S, 0.0001));
+        assertTrue(MatrixApproximatelyEquals(L_matrix_r, L, 0.0001));
```
Comment (tuhinkanti) [inline ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java]: Use assertThat instead of assertTrue. https://objectpartners.com/2013/09/18/the-benefits-of-using-assertthat-over-other-assert-methods-in-unit-tests/

File: ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+            int i = n % rows;
+            int j = (int) Math.floor(n / rows);
+            input2DArray[i][j] = x[n];
+        }
+        return input2DArray;
+    }
+
+    public boolean MatrixApproximatelyEquals(double[][] X, double[][] Y, double epsilon) {
+        boolean testOutput = true;
+        int printCnt = 0;
```
Comment (tuhinkanti) [inline ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/transform/RPCATest.java]: Please use a better and more expressive variable names for above following clean code principles

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	
+	private void initMatrices() {
+		this.L = MatrixUtils.createRealMatrix(this.X.getRowDimension(), this.X.getColumnDimension());
+		this.S = MatrixUtils.createRealMatrix(this.X.getRowDimension(), this.X.getColumnDimension());
+		this.E = MatrixUtils.createRealMatrix(this.X.getRowDimension(), this.X.getColumnDimension());
+	}
+	
+	private void computeRSVD() {
+		double mu = X.getColumnDimension() * X.getRowDimension() / (4 * l1norm(X.getData()));
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java]: Many of your mathematical operations might result in divide by 0 as you have no control over the input. Its a good practice to wrap them in try-catch blocks and do exception handling

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+		double objPrev = 0.5*Math.pow(X.getFrobeniusNorm(), 2);
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java]: Please use better variable names. I'd say from here on almost every variable name and method name should exactly communicate what it is for. You should find a lot of best practices online but try to describe behavior  in those names

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+			for (int j = 0; j < x[i].length; j++) {
+				l1norm += Math.abs(x[i][j]);
+			}
+		}
+		return l1norm;
+	}
+	
+	private double computeL(double mu) {
+		double LPenalty = lpenalty * mu;
+		SingularValueDecomposition svd = new SingularValueDecomposition(X.subtract(S));
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java]: :) use singularValueDecomposition instead of svd for the name. It is more expressive

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+			double l1Norm = computeL(mu);
+			double l2Norm = computeE();
+			
+			obj = computeObjective(nuclearNorm, l1Norm, l2Norm);
+			diff = Math.abs(objPrev - obj);
+			objPrev = obj;
+			
+			mu = computeDynamicMu();
+			
+			iter = iter + 1;
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/RPCA.java]: iter++ :)

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+        for (int i = 0; i < metricValues.length; i++) {
+            String value = completeDatapoints.get(timestamps[i]);
+            metricValues[i] = Double.parseDouble(value);
+        }
+        standardize(metricValues);
+
+        trainModel();
+        Metric predictions = predictAnomalies();
+        Metric predictionsNormalized = normalizePredictions(predictions);
+
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: Look up the design pattern Chain of responsibilities and see if you can apply it for lots of operations you chain

**Comment by tuhinkanti** [comment]
As a principle use exception handling generously than sparingly. You would not want the runtime to error out in any situation but instead handle errors gracefully

You might want to revisit the reason for inheritance. You can use composition or aggregation when the derived class isn't a specialization of the base class.

Great work though

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+        return input2DArray;
+    }
+
+    private double[] matrixToVector(double[][] matrix) {
+        int rows = matrix.length;
+        int cols = matrix[0].length;
+        double[] outputVector = new double[rows * cols];
+        for (int n = 0; n < outputVector.length; n++) {
+            int i = n % rows;
+            int j = (int) Math.floor(n / rows);
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionRPCATransform.java]: potential divide by 0 :)

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionTransform.java [hunk 25]
```diff
@@ -60,7 +60,7 @@
         Metric predictions = new Metric(getResultScopeName(), getResultMetricName());
         Map<Long, String> predictionDatapoints = new HashMap<>();
 
-        long detectionIntervalInSeconds = getDetectionIntervalInSeconds(constants.get(0));
+        long detectionIntervalInSeconds = getTimePeriodInSeconds(constants.get(0));
 
         //Create a sorted array of the metric's timestamps
         Map<Long, String> completeDatapoints = metrics.get(0).getDatapoints();
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionTransform.java]: I would be vary of writing code based on indices. This is prone to error when indices shift or data structures change. You are not protecting yourself to interface changes

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionTransform.java [hunk 26]
```diff
@@ -84,6 +84,21 @@
         throw new UnsupportedOperationException("This transform only supports anomaly detection on a single list of metrics");
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionTransform.java]: In most cases you don't want to throw explicit exceptions. Exceptions are very expensive as they unwind stacks and halt the runtime. This is the top google search and looks decent http://www.journaldev.com/1696/exception-handling-in-java

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionTransform.java [hunk 27]
```diff
@@ -84,6 +84,21 @@
...
     }
 
+    public long getTimePeriodInSeconds(String timePeriod) {
+        try {
+            //Parse constant for time period
+            String timeValueString = timePeriod.substring(0, timePeriod.length() - 1);
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionTransform.java]: extract timePeriod.length() - 1 to a variable and reuse

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionTransform.java [hunk 28]
```diff
@@ -84,6 +84,21 @@
...
+            String timeUnitString = timePeriod.substring(timePeriod.length() - 1);
```
Comment (tuhinkanti) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/transform/AnomalyDetectionTransform.java]: substring and indices are prone to error, is there a better way to get this

**Comment by shouvikmani** [comment]
@saaja-sfdc PR is ready to merge

Changes I made after code review:
- Exceptions for edge cases
- Code style and clarity

Changes I won't make due to scope of project and time constraints:
- Netflix RPCA code
- Netflix RPCA Tests

Changes I will make in a later PR (as they touch on larger design issues in all my anomaly detection transforms):
- Use Composition instead of Inheritance in certain cases

@tuhinkanti @afalko Thanks again for the help!
